### PR TITLE
Added support for namespace attributes

### PR DESCRIPTION
**Pull request by framp**
This PR adds support for namespace attributes, which is particularly useful for using svg with react.

The attributes are set via `setAttributeNS` using namespaces defined [here](http://www.w3.org/html/wg/drafts/html/master/syntax.html#adjust-foreign-attributes).
I've added `xmlns` on HTMLDOMProperty and all the other ones on SVGDOMProperty.

Useful svg-specific jsfiddle [here](http://jsfiddle.net/UVFBj/8/)

I hope this could be a first step toward supporting svg in react in a better way.

### CODE REVIEW

File: src/browser/ui/dom/SVGDOMPropertyConfig.js [hunk 1]
```diff
@@ -59,10 +66,34 @@ var SVGDOMPropertyConfig = {
     x1: MUST_USE_ATTRIBUTE,
     x2: MUST_USE_ATTRIBUTE,
     x: MUST_USE_ATTRIBUTE,
+    xlinkActuate: MUST_USE_ATTRIBUTE | HAS_NAMESPACE,
```
Comment (spicyj) [inline src/browser/ui/dom/SVGDOMPropertyConfig.js]: Seems like this HAS_NAMESPACE is redundant with the NamespaceProperties dict below -- is that right? If so, better to have only one of them.

File: src/browser/__tests__/ReactDOMSVG-test.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+/*jslint evil: true */
+
+'use strict';
+
+var React = require('React');
+var ReactTestUtils = require('ReactTestUtils');
+
+describe('ReactDOMSVG', function() {
+  it("allows a SVG element", function() {
+    var element = React.createElement('svg', {xmlnsXlink: 'http://www.w3.org/1999/xlink'});
```
Comment (spicyj) [inline src/browser/__tests__/ReactDOMSVG-test.js]: Would be nice if people didn't need to specify this by hand. I guess you need to when writing svg outside of React though?

Comment (spicyj) [inline src/browser/__tests__/ReactDOMSVG-test.js]: Based on this table, it sounds like it should be optional:

https://html.spec.whatwg.org/multipage/syntax.html#attributes-2:ascii-case-insensitive-2

This example seems to agree:

https://css-tricks.com/svg-use-external-source/

File: src/browser/ui/dom/SVGDOMPropertyConfig.js [hunk 3]
```diff
@@ -83,7 +114,18 @@ var SVGDOMPropertyConfig = {
     strokeOpacity: 'stroke-opacity',
     strokeWidth: 'stroke-width',
     textAnchor: 'text-anchor',
-    viewBox: 'viewBox'
+    viewBox: 'viewBox',
+    xlinkActuate: 'xlink:actuate',
```
Comment (spicyj) [inline src/browser/ui/dom/SVGDOMPropertyConfig.js]: If I'm understanding right, the name that you pass to setAttributeNS should actually be just 'actuate' here. When generating the markup, we do want to write xlink:actuate though…

**Comment by spicyj** [comment]
Did you test this in a browser to make sure that both the initial render and updates work properly? This PR is a lot closer than anything else I've seen so it would be great to iterate on this and get it to a point where it can be merged.

What do you think of the attribute names? Is it reasonable to have people write `xlinkActuate={...}` or is it worth changing the JSX transform to support `xlink:actuate={...}` or similar?

**Comment by framp** [comment]
Thanks for looking into this.

Yes, it's working in browsers and I'm using it on a project.
I'll be happy to work on it and improve it (and add more tests).

Regarding the attribute names - to me it's seems more consistent with the other react properties to just use camelCase.

File: src/browser/__tests__/ReactDOMSVG-test.js (hunk 2 above)
Comment (framp) [inline src/browser/__tests__/ReactDOMSVG-test.js]: Yes, this one is optional but other uses (`xlink:href` and `image` for example) are not

File: src/browser/ui/dom/SVGDOMPropertyConfig.js (hunk 3 above)
Comment (framp) [inline src/browser/ui/dom/SVGDOMPropertyConfig.js]: Luckily all the current browsers seems to ignore namespace set on the 2nd argument of setAttributeNS:

Testing it with the 'whatever' namespace.
http://jsfiddle.net/framp/yst5c721/1/
http://imgur.com/vXsHlGL

File: src/browser/__tests__/ReactDOMSVG-test.js (hunk 2 above)
Comment (spicyj) [inline src/browser/__tests__/ReactDOMSVG-test.js]: Let's not add xmlnsXlink then.

File: src/browser/ui/dom/SVGDOMPropertyConfig.js (hunk 3 above)
Comment (spicyj) [inline src/browser/ui/dom/SVGDOMPropertyConfig.js]: Okay, then this is fine as-is.

**Comment by spicyj** [comment]
Okay, cool. Can we combine HAS_NAMESPACE and NamespaceProperties then? (And nit: maybe call it PropertyNamespaces to reflect the meaning of the values?) After that I'll try to test this some but I think it should be good to merge.

**Comment by framp** [comment]
I merged `HAS_NAMESPACE` and `NamespaceProperties` in `PropertyNamespace` and `getAttributeNamespace` with `hasNamespace`.
I've also added more tests for the namespace attributes related to svg links.

Using the code in this [gist](https://gist.github.com/framp/a998d59e22ef036da83e), it should be easy to test the feature manually

File: src/browser/ui/dom/DOMProperty.js [hunk 4]
```diff
@@ -70,6 +70,8 @@ var DOMPropertyInjection = {
       );
     }
 
+    DOMProperty.getAttributeNamespace = domPropertyConfig.PropertyNamespaces;
```
Comment (spicyj) [inline src/browser/ui/dom/DOMProperty.js]: This overrides getAttributeNamespace each time injectDOMPropertyConfig is called, which means the HTML and SVG configs won't be merged properly. If you assign this in the loop below like the other lookup tables do then it should do the right thing.

File: src/browser/__tests__/ReactDOMSVG-test.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+              > 
+              <text xmlLang="en-US" xmlSpace="preserve" x="0" y="15" fill="black">React</text>
+            </a>
+        </g> 
+      </svg>
+    );
+    var svg = React.findDOMNode(instance);
+    var link = svg.childNodes[0].childNodes[0];
+    expect(link.tagName).toBe('a');
+    var href = link.getAttributeNS(SVGDOMNamespaces.xlink,'href');
```
Comment (spicyj) [inline src/browser/__tests__/ReactDOMSVG-test.js]: nit: space after comma here and for all of these

File: src/browser/__tests__/ReactDOMSVG-test.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    expect(href).toBe('http://facebook.github.io/react/');
+    var actuate = link.getAttributeNS(SVGDOMNamespaces.xlink,'actuate')
```
Comment (spicyj) [inline src/browser/__tests__/ReactDOMSVG-test.js]: (semicolon)

**Comment by framp** [comment]
Thanks, fixed the style issues and fixed the `getAttributeNamespace` problem

**Comment by framp** [comment]
Synced with master

**Comment by spicyj** [comment]
Thanks, merged in 1db70dbef1d6e15936bb2c93a5ca2400167ec9f2 with a few minor changes!

**Pull request closed without merging**